[metadata] Allow unresolvable/invalid class typerefs (#15767)

rbernon opened the pull request:
Still trying to fix L.A.Noire issues, I've implementing the change suggested in https://github.com/mono/mono/issues/15767#issuecomment-514831346 . I'm not entirely sure this is how it should be done, especially the error check removal from mono_class_setup_fields.

lambdageek commented:
> PRs welcome!

Oh.  Well I did say that, didn't I?

---

At first glance this looks reasonable, except I'd really love to have a test case.  Ideally there's already something in [`dotnet/runtime/src/tests`](https://github.com/dotnet/runtime/tree/master/src/tests).

I think something like the following IL test is good.

<details>
<pre>
// Metadata version: v4.0.30319
.assembly extern mscorlib
{
  .publickeytoken = (B7 7A 5C 56 19 34 E0 89 )                         // .z\V.4..
  .ver 4:0:0:0
}
.assembly extern ma // MISSING
{
  .ver 0:0:0:0
}
.assembly mm
{
  .hash algorithm 0x00008004
  .ver 0:0:0:0
}
.module mm.exe
.imagebase 0x00400000
.file alignment 0x00000200
.stackreserve 0x00100000
.subsystem 0x0003       // WINDOWS_CUI
.corflags 0x00000001    //  ILONLY


// =============== CLASS MEMBERS DECLARATION ===================

.class public auto ansi beforefieldinit mm.C
       extends [mscorlib]System.Object
{
  .field public int32 i
  .field public [ma]ma.Ma ma  // Field referencing missing assembly
  .method public hidebysig specialname rtspecialname
          instance void  .ctor() cil managed
  {
    // Code size       9 (0x9)
    .maxstack  8
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  nop
    IL_0007:  nop
    IL_0008:  ret
  } // end of method C::.ctor

} // end of class mm.C

.class private auto ansi beforefieldinit mm.Program
       extends [mscorlib]System.Object
{
    // public static void Main ()
    // {
    //   var x = new C ();
    //   x.i = 42;
    //   Console.WriteLine($"{x.i}");
    // }
    // Should create a new instance of C and print 42, even if C.ma field resolves to a missing class.
  .method private hidebysig static void  Main(string[] args) cil managed
  {
    .entrypoint
    // Code size       43 (0x2b)
    .maxstack  2
    .locals init (class mm.C V_0)
    IL_0000:  nop
    IL_0001:  newobj     instance void mm.C::.ctor()
    IL_0006:  stloc.0
    IL_0007:  ldloc.0
    IL_0008:  ldc.i4.s   42
    IL_000a:  stfld      int32 mm.C::i
    IL_000f:  ldstr      "{0}"
    IL_0014:  ldloc.0
    IL_0015:  ldfld      int32 mm.C::i
    IL_001a:  box        [mscorlib]System.Int32
    IL_001f:  call       string [mscorlib]System.String::Format(string,
                                                                object)
    IL_0024:  call       void [mscorlib]System.Console::WriteLine(string)
    IL_0029:  nop
    IL_002a:  ret
  } // end of method Program::Main

  .method public hidebysig specialname rtspecialname
          instance void  .ctor() cil managed
  {
    // Code size       8 (0x8)
    .maxstack  8
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  nop
    IL_0007:  ret
  } // end of method Program::.ctor

} // end of class mm.Program
</pre>
</details>


It would also be good to have a negative test that throws some exception when the missing field is actually needed for something.

lambdageek commented on mono/metadata/class-init.c:
```diff
@@ -403,6 +394,71 @@ mono_class_set_failure_and_error (MonoClass *klass, MonoError *error, const char
+	}
+
+	mono_metadata_decode_row (tt, tidx - 1, cols, MONO_TYPEREF_SIZE);
+
```
Maybe check that `cols [MONO_TYPEREF_SCOPE] & MONO_RESOLUTION_SCOPE_MASK == MONO_RESOLUTION_SCOPE_ASSEMBLYREF` - I'd rather be a bit conservative and only do this for missing references to another assembly, not for nested types or multi-module stuff.

lambdageek commented on mono/metadata/class-init.c:
```diff
@@ -403,6 +394,71 @@ mono_class_set_failure_and_error (MonoClass *klass, MonoError *error, const char
+	klass->interface_count = 0;
+	klass->interfaces_inited = 1;
+
+	/*g_print ("Load class %s\n", name);*/
```
```suggestion
	mono_trace(G_LOG_LEVEL_DEBUG, MONO_TRACE_TYPE, "Creating class '%s.%s' from invalid typeref", nspace, name);
```

lambdageek commented on mono/metadata/class-init.c:
```diff
@@ -403,6 +394,71 @@ mono_class_set_failure_and_error (MonoClass *klass, MonoError *error, const char
+	MONO_PROFILER_RAISE (class_loaded, (klass));
+
+	mono_class_set_type_load_failure (klass, "%s", mono_error_get_message (error));
+	error_init (error);
```
```suggestion
	mono_error_cleanup (error);
	error_init_reuse (error);
```

lambdageek reviewed (commented):
`mono_class_create_from_invalid_typeref` looks ok.

The biggest problem with this PR is that we have no tests.

Our checks of `mono_class_has_failure (klass)` or `mono_type_has_exceptions (type)` are not as pervasive as they could be.   So previously when we would fail early with a managed exception, now we might crash later because we didn't check for a failure and used some class assuming it's good.

It would be good to try a variety of missing typeref scenarios - missing locals, missing generic constraints, missing method parameters, etc.

rbernon commented:
I'll try to add some unit tests.

Also about the error handling, I wasn't sure how that should behave. The error is indeed pre-existing the call to `mono_class_create_from_invalid_typeref` and the idea is to make the type carry the error and clear it temporarily. However there were a few cases where `mono_class_create_from_invalid_typeref` itself could also fail. I thought about adding two MonoError parameters, one that is only supposed to be carried, and the other as usual for the call error itself.

lambdageek commented:
> However there were a few cases where `mono_class_create_from_invalid_typeref` itself could also fail. I thought about adding two MonoError parameters, one that is only supposed to be carried, and the other as usual for the call error itself.

My preference would be for two `MonoError`s. (I'm not sure the other Mono contributors necessarily agree - I think it's less confusing to use the error for just one thing at a time, but we're reused them in the past, so there's some precedent for just having one).  

You could, for example, use the last `MonoError *error` parameter for just when `mono_class_create_from_invalid_typeref` itself fails (like other functions that may fail); and a second one that carries just the failure we want to assign to the newly-created `klass`.  Something like:

```c
/** 
 * mono_class_create_from_invalid_typeref:
 *
 * Creates a new reference type in the given image from the given typeref token that is immediately
 * marked as failed to load with the \p class_failure error.
 *
 * On success returns the constructed (failed) class.  On failure sets \p error and returns NULL.
 *
 * LOCKING: takes the loader lock
 */
MonoClass *
mono_class_create_from_invalid_typeref (MonoImage *image, uint32_t token, /*const*/ MonoError *class_failure, MonoError *error);
```
(If you can actually make it `const` that would be great, but I'm not sure if our internal APIs are good enough to work with a `const MonoError *`)

lambdageek commented:
We should also find out if this is good enough for compound types like arrays.  I can't tell if the current version can avoid failures when faced with:
```csharp
  class GoodClass {
    private MissingClass[] badField;
  }
```

It kind of depends on whether we look at `mono_class_has_failure` when making an array.

rbernon commented:
So I'm adding the following test case:

```csharp
using System;

class H<T>
{
}

class C
{
	public M missing;
	public M[] missing_array;
	// public M.N missing_nested;
	// public G<M> missing_generic;
	public H<M> missing_generic_param;
	public void bar(M missing) {}

	public int field;
	public C()
	{
		field = 42;
	}
}

class D
{
	public M missing;
	public D()
	{
		missing = null;
		missing.foo(0);
	}
}

class T
{
	public M missing;
	public static void bar(C c, M missing) { c.bar(missing); }

	static int Main (string [] args)
	{
		C c = new C();
		try
		{
			bar(c, null);
			return 1;
		}
		catch (System.IO.FileNotFoundException e)
		{
			// expected
			Console.WriteLine("bar failed");
		}
		try
		{
			D d = new D();
			return 2;
		}
		catch (System.IO.FileNotFoundException e)
		{
			// expected
			Console.WriteLine("new D() failed");
		}
		return 0;
	}
}
```

With the following missing assembly:
```csharp
public class M
{
	public class N {}
}
public class G<T>
{
}
```

The test passes with .NET Framework and not with current Mono of course, but it also passes *too much* with the PR: none of the two expected exceptions are raised. It's not obvious to me where the `null` assignment is handled, as it seems like it is where we miss a class failure check.

Also I think I'm also going to leave the generic class and nested class cases aside, as commented:
1) The generic class requires supporting invalid generic typerefs, which I'm not sure we can easily create.
2) The nested class case is explicitly left unimplemented as per one of the remarks above, but probably would depend as well of similar things.

marek-safar commented:
@monojenkins approve

marek-safar commented:
@lambdageek what are the next steps here?

lambdageek commented:
@monojenkins build

madewokherd commented:
> The test passes with .NET Framework and not with current Mono of course, but it also passes _too much_ with the PR: none of the two expected exceptions are raised. It's not obvious to me where the `null` assignment is handled, as it seems like it is where we miss a class failure check.

The call should probably be checked somewhere around when check_call_signature is called in method-to-ir.c's MONO_CEE_CALL case, but not in check_call_signature itself because we don't want the JIT to fail.

The field assignment would be handled by the MONO_CEE_STFLD case. I guess it checks for failures loading the type containing the field, but not the type of the field itself.

rbernon commented:
Thanks! I could make the two problematic cases throw exceptions as expected, and I updated the MR with the latest revision. It should address the few comments above and it includes a test case compiled from the following source:

**bug-15767.cs**:
```cs
using System;

class H<T>
{
}

class C
{
	public M missing;
	public M[] missing_array;
	// public M.N missing_nested;
	// public G<M> missing_generic;
	public H<M> missing_generic_param;
	public void bar(M missing) {}

	public int field;
	public C()
	{
		field = 42;
	}
}

class D
{
	public M missing;
	public D()
	{
		missing = null;
	}
}

class T
{
	public M missing;
	public static void bar(C c, M missing) { c.bar(missing); }

	static int Main (string [] args)
	{
		C c = new C();
		try
		{
			D d = new D();
			return 2;
		}
		// .NET throws System.IO.FileNotFoundException
		// Mono throws System.TypeLoadexception
		catch (System.Exception e)
		{
			// expected
		}
		try
		{
			bar(c, null);
			return 1;
		}
		// .NET throws System.IO.FileNotFoundException
		// Mono throws System.TypeLoadexception
		catch (System.Exception e)
		{
			// expected
		}
		return 0;
	}
}
```

**bug-15757-missing.cs**:
```cs
public class M
{
	public class N {}
}
public class G<T>
{
}
```

madewokherd commented:
This breaks the bug-81691 test. I guess we need some validity check when getting the field type through reflection.

madewokherd commented:
This breaks the bug-81691 test, I guess we need to check that the type is valid when getting the field type through reflection.

madewokherd commented:
I think adding a check in ves_icall_RuntimeFieldInfo_ResolveType would fix the bug-81691 test, but I wonder if mono_type_get_object_handle should actually be changed (so we can't get a Type object for any unresolvable type).

rbernon committed:
[metadata] Allow unresolvable/invalid class typerefs (#15767)

Until they are actually used. But in the case they aren't, they should
not prevent an assembly to load.

rbernon commented:
I added a check for class failures in mono_type_get_object_checked, which should make the `bug-81691` test pass again.

rbernon commented:
The additional checks caused some more tests to fail, and some were missing that I missed. I've tweaked it a bit and hopefully all the tests should pass now. It's a bit ad-hoc though...

rbernon commented:
Another update to make that last test pass. I've added a few helpers to make the failure checks cleaner.

I also changed the invalid typeref class to keep the original error instead of changing it to a type load error, so that we then correctly throw FNFE later instead of TypeLoadException.

Also made its parent the `mono_defaults.object_class` as otherwise it would create a type load failure itself.

rbernon commented:
So it fails to throw the exceptions when AOT is used, but I don't really where the `mono-ci (Linux x64)` failure comes from...

rbernon commented:
Actually it was missing some checks in interpreter mode too, I added them.

However, I can also see some crashes in the AOT compilation and I'm not sure what to do about it though.

The problem is that it's not able to execute the exception constructor when running `mono_method_to_ir` during some of the test case compilation. It looks like it's not the only place where exceptions are created, so I suspect these case don't work either with AOT, but probably they aren't present in the tests.
